### PR TITLE
Assignment 7/8 Conner Hanson + Trevor Stewart

### PR DESCRIPTION
**Pull request by connerhanson**

### CODE REVIEW

**Comment by SergeiAshikhmin** [comment]
I will review this :)

File: Assignment.Tests/SampleDataTests.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        Assert.AreEqual(1, names.Count());
+        Assert.AreEqual(("Issiah", "Bester"), names.Single());
+        
+    }
+
+    [TestMethod]
+    public void FilterByEmailAddress_FilterUsesContains_Success()
+    {
+        //An example of what the filter could be
+        Predicate<string> filter = s => s.Contains(".gov");
```
Comment (SergeiAshikhmin) [inline Assignment.Tests/SampleDataTests.cs:133]: Great test 👍

**Review by SergeiAshikhmin — COMMENTED** [review]
Good job 👍

Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔
Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string. ✔

Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔
Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. (Sort the addresses first then select). ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔

Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔
Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.

Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌

Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ❌
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ✔
Choose simplicity over complexity ✔

**Comment by HadiM2** [comment]
Reviewing

File: Assignment.Tests/SampleDataTests.cs (hunk 1 above)
Comment (Tvor369) [reply Assignment.Tests/SampleDataTests.cs:133]: thank you :)

File: Assignment.Tests/NodeTests.cs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+
+        var result = pets.Select(node => node.Data).Where(data => data.Contains('i'));
+        List<string> expected = ["fish", "bird"];
+
+        Assert.IsTrue(expected.SequenceEqual(result));
+    }
+
+
+    [TestMethod]
+    public void ChildItems_ReturnMaximumOfThree_Success()
```
Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:57]: There could be a good argument made for whether maximum should be inclusive or exclusive.

Comment (BenjaminMichaelis) [inline Assignment.Tests/NodeTests.cs:57]: What about edge cases? What if I give it a max of `-4`? Should it throw an exception, should it return empty? Good things to think about.

File: Assignment/Node.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        //We want the remaining items, so exclude the start node
+        Node<T> cur = Next;
+        int count = 0;
+
+        while (cur.Next != this && count < maximum)
+        {
+            yield return cur;
+            cur = cur.Next;
+            count++;
+        }
```
Comment (BenjaminMichaelis) [inline Assignment/Node.cs:92]: Why not just use linq to simplify

File: Assignment/SampleData.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    }
+
+    // 6.
+    public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+    {
+        var uniqueSortedList = people
+            .Select(person => person.Address.State)
+            .Distinct()
+            .OrderBy(state => state, StringComparer.OrdinalIgnoreCase);
+        string result = uniqueSortedList.Aggregate((concat, str) => $"{concat}, {str}");
```
Comment (BenjaminMichaelis) [inline Assignment/SampleData.cs:79]: Why not use string.Join()?
